File: javascript/connect/connect.js
//
// This is only a SKELETON file for the 'Connect' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Board {
  constructor(board) {
    this.board = board;
    this.height = board.length;
    this.width = board.length > 0 ? board[0].split(' ').length : 0;
    
    // Parse the board into a 2D grid for easier manipulation
    this.grid = this.parseBoard();
  }

  parseBoard() {
    const grid = [];
    
    for (let row = 0; row < this.height; row++) {
      const rowCells = this.board[row].trim().split(' ');
      grid.push(rowCells);
    }
    
    return grid;
  }

  winner() {
    // Check if player X has won (connected left to right)
    if (this.hasPlayerWon('X')) {
      return 'X';
    }
    
    // Check if player O has won (connected top to bottom)
    if (this.hasPlayerWon('O')) {
      return 'O';
    }
    
    // No winner
    return '';
  }

  hasPlayerWon(player) {
    if (player === 'X') {
      // Player X wins by connecting left to right
      // Find all X cells on the left edge
      const startCells = [];
      for (let row = 0; row < this.height; row++) {
        if (this.grid[row][0] === 'X') {
          startCells.push([row, 0]);
        }
      }
      
      // Check if any of these cells can reach the right edge
      for (const [row, col] of startCells) {
        const visited = new Set();
        if (this.dfs(row, col, 'X', visited, (r, c) => c === this.grid[r].length - 1)) {
          return true;
        }
      }
    } else if (player === 'O') {
      // Player O wins by connecting top to bottom
      // Find all O cells on the top edge
      const startCells = [];
      if (this.grid.length > 0) {
        for (let col = 0; col < this.grid[0].length; col++) {
          if (this.grid[0][col] === 'O') {
            startCells.push([0, col]);
          }
        }
      }
      
      // Check if any of these cells can reach the bottom edge
      for (const [row, col] of startCells) {
        const visited = new Set();
        if (this.dfs(row, col, 'O', visited, (r, c) => r === this.height - 1)) {
          return true;
        }
      }
    }
    
    return false;
  }

  dfs(row, col, player, visited, isDestination) {
    // Create a unique key for the cell
    const key = `${row},${col}`;
    
    // If we've already visited this cell, return false
    if (visited.has(key)) {
      return false;
    }
    
    // Mark the cell as visited
    visited.add(key);
    
    // If this cell is a destination, return true
    if (isDestination(row, col)) {
      return true;
    }
    
    // Get all adjacent cells
    const adjacentCells = this.getAdjacentCells(row, col);
    
    // Check if any adjacent cell can reach the destination
    for (const [nextRow, nextCol] of adjacentCells) {
      // Only consider cells that contain the player's stone
      if (this.isValidCell(nextRow, nextCol) && this.grid[nextRow][nextCol] === player) {
        if (this.dfs(nextRow, nextCol, player, visited, isDestination)) {
          return true;
        }
      }
    }
    
    return false;
  }

  getAdjacentCells(row, col) {
    // In a hexagonal grid, each cell has 6 adjacent cells
    // The exact directions depend on the representation
    // Based on the test cases, it seems like the adjacent cells are:
    // (row-1, col), (row-1, col+1), (row, col-1), (row, col+1), (row+1, col-1), (row+1, col)
    return [
      [row - 1, col],     // top
      [row - 1, col + 1], // top-right
      [row, col - 1],     // left
      [row, col + 1],     // right
      [row + 1, col - 1], // bottom-left
      [row + 1, col]      // bottom
    ];
  }

  isValidCell(row, col) {
    return row >= 0 && row < this.height && col >= 0 && col < this.grid[row].length;
  }
}

File: python/connect/connect.py
class ConnectGame:
    def __init__(self, board):
        # Parse the board string into a 2D grid
        lines = board.strip().split('\n')
        self.board = []
        
        for line in lines:
            # Remove leading/trailing whitespace and split by spaces
            row = [cell for cell in line.strip().split(' ') if cell]
            self.board.append(row)
        
        self.height = len(self.board)
        self.width = max(len(row) for row in self.board) if self.board else 0

    def get_winner(self):
        # Check if player O has won (connected top to bottom)
        if self._has_o_won():
            return "O"
        
        # Check if player X has won (connected left to right)
        if self._has_x_won():
            return "X"
        
        # No winner
        return ""
    
    def _has_x_won(self):
        """Check if player X has won by connecting left to right."""
        if not self.board:
            return False
        
        # Find all X stones on the left edge
        left_edge_x = []
        for i in range(self.height):
            if self.board[i] and self.board[i][0] == "X":
                left_edge_x.append((i, 0))
        
        # For each X on the left edge, check if it's connected to the right edge
        for start in left_edge_x:
            visited = set()
            if self._dfs_x(start[0], start[1], visited):
                return True
        
        return False
    
    def _has_o_won(self):
        """Check if player O has won by connecting top to bottom."""
        if not self.board:
            return False
        
        # Find all O stones on the top edge
        top_edge_o = []
        if self.board[0]:
            for j in range(len(self.board[0])):
                if self.board[0][j] == "O":
                    top_edge_o.append((0, j))
        
        # For each O on the top edge, check if it's connected to the bottom edge
        for start in top_edge_o:
            visited = set()
            if self._dfs_o(start[0], start[1], visited):
                return True
        
        return False
    
    def _is_valid_position(self, i, j):
        """Check if position (i, j) is valid on the board."""
        return (0 <= i < self.height and 
                0 <= j < len(self.board[i]))
    
    def _get_neighbors(self, i, j):
        """Get valid neighboring positions for position (i, j)."""
        # In a hexagonal grid with this representation, the neighbors are:
        # The grid is offset, so the neighbors depend on the row
        potential_neighbors = [
            (i-1, j), (i-1, j+1),  # Up-left, Up-right
            (i, j-1), (i, j+1),    # Left, Right
            (i+1, j-1), (i+1, j)   # Down-left, Down-right
        ]
        
        # Filter out invalid positions
        return [(ni, nj) for ni, nj in potential_neighbors if self._is_valid_position(ni, nj)]
    
    def _dfs_x(self, i, j, visited):
        """DFS to check if X at position (i, j) is connected to the right edge."""
        # Check if position is valid and is an X
        if not self._is_valid_position(i, j) or self.board[i][j] != "X":
            return False
        
        # Check if already visited
        if (i, j) in visited:
            return False
        
        # Mark as visited
        visited.add((i, j))
        
        # Check if reached the right edge
        if j == len(self.board[i]) - 1:
            return True
        
        # Check all valid neighbors
        for ni, nj in self._get_neighbors(i, j):
            if self._dfs_x(ni, nj, visited):
                return True
        
        return False
    
    def _dfs_o(self, i, j, visited):
        """DFS to check if O at position (i, j) is connected to the bottom edge."""
        # Check if position is valid and is an O
        if not self._is_valid_position(i, j) or self.board[i][j] != "O":
            return False
        
        # Check if already visited
        if (i, j) in visited:
            return False
        
        # Mark as visited
        visited.add((i, j))
        
        # Check if reached the bottom edge
        if i == self.height - 1:
            return True
        
        # Check all valid neighbors
        for ni, nj in self._get_neighbors(i, j):
            if self._dfs_o(ni, nj, visited):
                return True
        
        return False
